### PR TITLE
Support for max_window_layers

### LINKED ISSUE #147: [feat] Support `max_window_layers`
# 🎯 **Goal (What & Why)**
Enable support for `max_window_layers` in Qwen2 models. Qwen2 uses sliding window attention only in the highest layers, requiring a `max_window_layers` parameter to configure it. This is a blocker for loading and training Qwen2 models in Fast-LLM, cf. #135.

# 🚀 **Execution Plan**
### **Step 1: Minimal Working Version (Must-Have)**
- Add `max_window_layers` to Fast-LLM's model config as in [Qwen2's config.json](https://huggingface.co/Qwen/Qwen2-7B/blob/453ed1575b739b5b03ce3758b23befdb0967f40e/config.json#L13).
- Modify attention layers to respect `max_window_layers` setting, following [`modular_qwen2.py`](https://github.com/huggingface/transformers/blob/8eaae6bee973e2068b5ed5622a708d99ce3dd862/src/transformers/models/qwen2/modular_qwen2.py#L72)'s example.

### **Step 2: Optional Extensions (Future Work)**
* Investigate whether or not important models use sliding window strategies that go beyond what is implemented in Step 1.
* Explore whether a more generalized `sliding_window_type` setting is needed.
* Consider refactoring if and only if other architectures require different attention behaviours.

**Step 2 is optional and should be discussed separately after Step 1 is implemented.**  

# 📌 **Acceptance Criteria** (Must-Haves for Completion)
* The feature must be **functional and tested**. 
* The new configuration option must be **documented**.
* The PR must include a **performance/impact summary**.  
* **No refactors unless directly necessary** for this feature.  

# 🛠️ **Project Management**
- [x] **Assign the project to the Fast-LLM project.**
- [x] **Set the `Estimate` field (in days) in the GitHub project.**
- [x] **Use the `Size` field to categorize the PR size (Small/Medium/Large).**
- [x] **Assign an owner when opening the issue.**

### PR DESCRIPTION
**Pull request by bigximik**
# ✨ Description

Closes #147
Also, added assert for `window_size` usage for non flash attention

## 🔍 Type of change

Select all that apply:

- [x] 🐛 **Bug fix** (non-breaking change that addresses a specific issue)
- [x] 🚀 **New feature** (non-breaking change that adds functionality)
- [ ] ⚠️ **Breaking change** (a change that could affect existing functionality)
- [ ] 📈 **Performance improvement/optimization** (improves speed, memory usage, or efficiency)
- [ ] 🛠️ **Code refactor** (non-functional changes that improve code readability, structure, etc.)
- [ ] 📦 **Dependency bump** (updates dependencies, including Dockerfile or package changes)
- [ ] 📝 **Documentation change** (updates documentation, including new content or typo fixes)
- [ ] 🔧 **Infrastructure/Build change** (affects build process, CI/CD, or dependencies)

## 📝 Changes

List the key changes introduced in this PR:

1. Adds support for `max_window_layers` a threshold on which layers to use sliding window attention
2. Added assert for `window_size` usage for non flash attention

## ✅ Checklist

Make sure the following tasks are completed before submitting the PR:

### General

- [x] 📜 I have read and followed the [contributing guidelines](https://servicenow.github.io/Fast-LLM/developers/contributing).
- [x] 🏷️ I am using a clear and descriptive PR title that summarizes the key change or feature introduced.
- [x] 🎉 The functionality is complete, and I have tested the changes. **(tested only new test cases)**
- [ ] 📝 I have updated the documentation if needed. **(not applicable)**
- [x] ⚠️ The change does not introduce any new issues (e.g., runtime warnings, type checker errors, linting problems, unhandled edge cases).
- [x] 🧩 I have commented my code, especially in hard-to-understand areas.

### Testing

- [x] 🧪 I have added or updated tests to cover my changes.
- [x] ✔️ New and existing tests pass locally with my changes. **(tested only new test cases)**
- [ ] 🚦 I have tested these changes on GPUs and verified training stability. **(not applicable)**
- [ ] 🏋️ I have tested the changes on realistic training workloads, if applicable. **(not applicable)**

### CODE REVIEW

**Comment by jlamypoirier** [comment]
@bigximik Why did you drop support for non-flash windowed attention? It should be supported.

File: fast_llm/layers/transformer/attention.py [hunk 1]
```diff
@@ -274,6 +274,20 @@ def _query_key_value_backward(
...
 
+    
+    def _decide_window_size(self) -> int | None:
+        # NOTE: This is a temporal solution for qwen 2.X
+        # https://github.com/huggingface/transformers/blob/5e2183f344911aa82aba0b83778a4f196cff378e/src/transformers/models/qwen2/modular_qwen2.py#L71
+        # TODO: make universal per layer config
+        window_size = self._config.window_size
+        if (
+            self._config.max_window_layers is not None
+            and self._layer_index < self._config.max_window_layers
```
Comment (jlamypoirier) [inline fast_llm/layers/transformer/attention.py:285]: I think this is incorrect because layer index starts at 1 for some reason.

**Comment by tscholak** [comment]
Hi @jlamypoirier, I strongly prefer that we don't merge unfinished, experimental features just to "play with them", especially when they're complex and introduce long-term maintenance overhead.

Let me reiterate that:

1. This feature is not urgent. The immediate priority is LoRA, and we shouldn't add distractions. LoRA as it is scoped right now also doesn't depend on this, contrary to what #180 suggests. I just commented on that.
2. The fact that this needs an experimental flag tells me it's not ready to be merged. If it's still evolving, it should stay in this branch until it's ready.
3. The `max_window_layers` discussion from #157 isn't urgent either. Qwen2 defines it but published models don't actually use it, so we don't need to rush a general solution.
4. There will be a real need for this feature, when we experiment with SSM-transformer hybrid configurations, but that is weeks away.

Please let's keep focus on the roadmap and merge features when they're fully ready and actually needed. Thanks.
